hyperspace: init at 1.1.4

dtzWill opened the pull request:
###### Motivation for this change

Mastodon client :).
(electron-based)

FWIW, the license is new to us and not widely used.

Added it to the list, but am unsure how it should be handled.
Hopefully it's alright to include software under this license,
but I mention it because I would appreciate a second opinion O:).

Fixes #57357.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

vyp commented on lib/licenses.nix:
```diff
@@ -541,6 +541,11 @@ lib.mapAttrs (n: v: v // { shortName = n; }) {
 
+  npl1 = {
+    url = "https://git.pixie.town/thufie/NPL";
+    fullName = "Non-Violent Public License v1";
```
```suggestion
    fullName = "Non-Violent Public License v1";
    free = false;
```

From the URL:

> A Note to those concerned about "Free Software".
>
> The Non-Violent Public License intentionally does not meet the Free Software Definition because of the restrictions put in place to protect others from violence aided by the use of the copyrighted work. "Freedom 0" is explicitly violated to afford these protections and prevent unethical usage of the program to actually Free Society. This license does preserve freedoms 1-3 and can be referred to as a "three freedoms" license. Similarly to the GNU AGPL this license has the additional requirement that server-side source code modifications to public-facing server software must be published for users as well as the usual expectations for FLOSS software with the exception of restrictions on Freedom 0 for ethical purposes.

So this is nonfree software by definition, but still, should we list it as nonfree when it still gives 3 out of the 4 freedoms for free software? I don't know but I'll try to elaborate a bit.

I respect the spirit of the license itself, it does seem to mean well but I guess the most 'controversial' part (for me) is its absolute pacifist nature i.e. no bodily harm/war/incarceration would imply not even allowed for self defense.

But without even getting into any discussion of ethics, I think really what would happen in practice with respect to violations of this license is that a state would likely already have their own laws about the various restrictions in the "Ethical Restrictions on usage" section, and that the state's laws would be favored over copyright law (which is what this license falls under). So really I think this entire "Ethical Restrictions" section is probably useless (i.e. seems like the wrong domain to be in a copyright license), but I'm no lawyer and maybe law in general doesn't work like that.

But if it is a 'useless' section, does that equate to actually having [freedom 0](https://www.gnu.org/philosophy/free-sw.en.html) with this license? Meaning it would be acceptable to **not** have `free = false`? This is what I'm unsure about but I guess either way it's probably just safer to have `free = false`. (And if we do so, maybe we should also give it `unfreeRedistributable` license?)

---

PS. I wonder what free society really means to the author of the license as I've seen these absolute pacifist views in free software/culture communities before but usually they are also very anti-copyleft unlike here (for similar reasons like anti-incarceration, seeing copyleft as inherently 'violent', etc). Also the "Extra Restriction in the CNPL" seems like an anti-capitalist restriction, whereas I've always thought free society was more pro capitalism. But I guess I'm probably remembering a different idea of free society than the author of this license.

Also the license probably needs to define a lot more terms it uses to really be usable but again, I'm no lawyer so can't really suggest anything without knowing what exactly.

jonringer replied:
also, a license addition should be done in a separate commit

alyssais replied:
> So this is nonfree software by definition, but still, should we list it as nonfree when it still gives 3 out of the 4 freedoms for free software?

It should be listed as non-free, IMO. It’s authors clearly do not intend for it to be free, and it’s not up to us to decide that part of the license is meaningless and should be ignored.

jonringer commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,75 @@
+    alsaLib
+  ];
+
+  runtimeDependencies = [ libGL udev.lib ];
```
didn't know about this, nice :)

dtzWill replied:
Apparently it's documented but I didn't either! Found it in other expressions when trying to fix some issue while packaging haha :).

matthiasbeyer commented:
The software works, would be nice to have it in nixpkgs...

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,74 @@
+{ stdenv, fetchurl, lib
```
Please only input stdenv or lib.

dtzWill replied:
Thanks.  In light of recent RFC or whatever on not using stdenv.lib I think this is preferred now so will leave as-is.

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,74 @@
+, autoPatchelfHook, dpkg
+, libX11, libXext, libXi, libXau, libXrender, libXft, libXmu, libSM, libXcomposite, libXfixes, libXpm
+, libXinerama, libXdamage, libICE, libXtst, libXaw, fontconfig, pango, cairo, glib, libxml2, atk, gtk3
+, gdk-pixbuf
+# more
+, nss, nspr
+, libXScrnSaver
+, alsaLib
+# rt?
+, udev, libGL
```
```suggestion
, gdk-pixbuf, nss, nspr, libXScrnSaver, alsaLib, udev, libGL
```

dtzWill replied:
d'oh sorry for leaving these in there! I do like grouping sometimes for maintainability but these are absolutely useless categories haha.  Anyway will fix.

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,74 @@
+
+stdenv.mkDerivation rec {
+  pname = "hyperspace";
+  version = "1.0.3";
+  src = fetchurl {
```
```suggestion
  version = "1.0.3";

  src = fetchurl {
```

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,74 @@
+  '';
+
+  installPhase = ''
+    mkdir -p $out
+
+    # /usr/share
+    mv ./usr/* $out/
+
+    # /opt/Hyperspace\ Desktop
+    mv ./opt/Hyperspace\ Desktop $out/share/hyperspace
+
+    mkdir -p $out/bin
+    ln -s $out/share/hyperspace/hyperspace $out/bin/hyperspace
+
+    rm -vrf $out/share/hyperspace/{libGLESv2.so,libEGL.so,swiftshader}
+
+    substituteInPlace $out/share/applications/hyperspace.desktop \
+     --replace 'Exec="/opt/Hyperspace Desktop/hyperspace"' \
+               'Exec="$out/bin/hyperspace"'
```
```suggestion
    mkdir -p $out

    mv ./usr/* $out/
    mv ./opt/Hyperspace\ Desktop $out/share/hyperspace

    mkdir -p $out/bin
    ln -s $out/share/hyperspace/hyperspace $out/bin/hyperspace

    rm -vrf $out/share/hyperspace/{libGLESv2.so,libEGL.so,swiftshader}

    substituteInPlace $out/share/applications/hyperspace.desktop \
     --replace 'Exec="/opt/Hyperspace Desktop/hyperspace"' \
               'Exec="$out/bin/hyperspace"'
```

dtzWill replied:
Fixed (looks like comments were unwanted, sounds good!).

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,74 @@
+    # From maxx, only w/gtk3
+    stdenv.cc.cc libX11 libXext libXi libXau libXrender libXft libXmu libSM libXcomposite libXfixes libXpm
+    libXinerama libXdamage libICE libXtst libXaw fontconfig pango cairo glib libxml2 atk gtk3
+    gdk-pixbuf
+    # more
+    nss nspr
+    libXScrnSaver
+    alsaLib
```
```suggestion
    gdk-pixbuf nss nspr libXScrnSaver alsaLib
```

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,74 @@
+  '';
+
+  buildInputs = [
+    # From maxx, only w/gtk3
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,74 @@
+    license = licenses.npl1;
+    maintainers = with maintainers; [ dtzWill ];
+  };
+}
+
```
```suggestion
}
```

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -25557,4 +25557,5 @@ in
 
   gortr = callPackage ../servers/gortr {};
 
+  hyperspace = callPackage ../applications/misc/hyperspace { };
```
Please don't add it at the end but under a fitting headline and alphabetically sorted.

dtzWill committed:
licenses: add npl4, "Non-Violent Public License v4"

dtzWill committed:
hyperspace: init at 1.1.4

dtzWill commented:
Updates/notes for anyone reading this in the future:

* NPL is a bit of a mess, it seems? Not free (I think we discussed this somewhere previously too), according to hyperspace discussion on changing its license (not linking to avoiding bombing them with mention, LMK if curious/can't find), bumped to v4 from the v1 included with this PR originally.  Authoritative URL for this version is tricky, it seems NPL author wants folks using v7+ and deprecated the v4 repository ... anyway this seems correct for hyperspace, which explicitly uses "NPLv4".
* As I recall, the runtime deps are needed because they're looked for but not linked against explicitly, now that we have to pull in mesa for libgbm I'm unsure if that makes sense re:libGL.  Leaving as I think libGL is preferred unless mesa itself is explicitly needed.
* There's an appimage alternative they provide which might be preferred (if simpler to package), I don't know if there's a reason I went for the .deb.  Seems less opaque :).  In theory we could maybe bundle this ourselves but I haven't explored this, as the tooling isn't very Nix-compat (node).  Staying the course for now.

---

Addressed review feedback, thank you! Sorry for delay! :)

----

Tested the result and seems to work as expected so far.

The icon is... X11 logo? Which suggests something strange happening but I don't think is critical.

--

Well anyway, here's the latest hyperspace :).

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,61 @@
+  version = "1.1.4";
+
+  src = fetchurl {
+    url = "https://github.com/hyperspacedev/${pname}/releases/download/v${version}/${pname}_${version}_amd64.deb";
```
```suggestion
    url = "https://github.com/hyperspacedev/hyperspace/releases/download/v${version}/hyperspace_${version}_amd64.deb";
```
easy to visit links are important if you want to research with one click if this can be build from source.

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,61 @@
+  runtimeDependencies = [ libGL udev ];
+
+  dontBuild = true;
+  dontStrip = true;
+  dontPatchELF = true;
```
Those break the binary?

SuperSandro2000 commented on pkgs/applications/misc/hyperspace/default.nix:
```diff
@@ -0,0 +1,61 @@
+    description = "A beautiful, fluffy client for the fediverse";
+    homepage = "https://hyperspace.marquiskurt.net";
+    license = licenses.npl4;
+    maintainers = with maintainers; [ dtzWill ];
```
Presumable platforms.linux, right?

onny commented:
Please rebase

Mic92 commented:
Closing because of lack of activity from the author. Please re-open if you plan to work on this.
